src/transformStateWithClones.js

```diff
@@ -7,7 +7,32 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const CloneObject = { ...state };
+  const result = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(CloneObject, action.extraData);
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete CloneObject[key];
+        }
+        break;
+
+      case 'clear':
+        for (const key in CloneObject) {
+          delete CloneObject[key];
+        }
+        break;
+    }
+
+    result.push({ ...CloneObject });
```

Comment:
Pushing the copy of the object to the array should be done at the end of each loop cycle but outside of the 'switch' block, after the current action is processed. This ensures that the same copy is not mutated by subsequent actions.